### ISSUE TITLE
NPC

### ISSUE THREAD

**Issue by 4mitabh** [RafaelBarbosatec/bonfire#201]
Thank you! Bonfire is a great help for anyone writing RPGs.

I see that you have players and enemies as part of the engine. Is there a plan to add NPCs a well?
NPCs are central to interaction in RPG style games.

If now what is the recommended way of modelling NPCs.

**Comment by 4mitabh** [comment]
I see that https://github.com/RafaelBarbosatec/darkness_dungeon/tree/master/lib/npc has NPCs extended from GameDecoration. This works... but maybe it's a good idea to add the base class for NPCs in bonfire itself.

**Comment by RafaelBarbosatec** [comment]
Yeah, you can use GameDecoration or Enemy without movements. 😊
And use "seePlayer" or "seeCompoment" to interact with other components.
I don't know if I would need to create something specific for NPC. But I will be analyzing.

**Comment by 4mitabh** [comment]
I was thinking, NPCs should be the base class for Enemy. The NPC class can have all the functionality related to 'seeing', 'saying' and 'moving'. The Enemy class can extend it with attacks, health etc.

**Comment by 4mitabh** [comment]
```
                         ┌─────────────┐
                         │             │
                         │     NPC     │
                         │             │
                         └──────┬──────┘
                                │
       ┌────────────────────────┼───────────────────────┐
       │                        │                       │
┌──────▼──────┐          ┌──────▼──────┐         ┌──────▼──────┐
│             │          │             │         │             │
│   Enemy     │          │   Neutral   │         │    Ally     │
│             │          │             │         │             │
└─────────────┘          └─────────────┘         └─────────────┘

```

If you agree with the idea, I can submit a PR after refactoring the Enemy class into a NPC base class, and a derived Enemy class. Users of the package can implement their NeutralNPCs and AllyNPCs derived from the NPC class.

ps: I'm excited to see Bonfire mentioned in the official Google blog while talking about Flame 1.0 release.
Good to see your awesome work get recognised.

**Comment by RafaelBarbosatec** [comment]
So I still don't think I would need to create an element called NPC. Because if we start going down that path, everything that's kind of elements in an RPG game, we're going to have to create a type, but that's not the idea. The idea is to have a few core elements and everything else to be created from them.

**Comment by 4mitabh** [comment]
> The idea is to have a few core elements and everything else to be created from them.

That design principle makes a lot of sense.

I'm still struggling with what to do with NPCs... Deriving it from wall decorations does not seen right, plus wall decorations do not move. Extending enemies instead is a better idea, but does not seem like a clean approach either. 

I see that you're working on other areas of the engine right now, do review the NPC design when you a chance.

**Comment by RafaelBarbosatec** [comment]
I got it.
But is not necessary extending from "Enemy" to give movements to component, you can use a simple GameComponent and adds "Movement" mixin. If you can recieve damage you can use "Attackable" mixin and so on. We try delivery the more modular possible using [mixins](https://bonfire-engine.github.io/#/mixins). This way you can build a component according to your need.

But it’s a good point. I'll be thinking better
Sorry for late to answer. I was on vacation.

**Comment by RafaelBarbosatec** [comment]
Hi @4mitabh !
Further analyzing this matter. I see it makes complete sense. If still available to submit a PR. will be very welcome.

**Comment by 4mitabh** [comment]
Great!
Let me take a stab at it.

**Comment by 4mitabh** [comment]
So, I thought some more about it, and also looked at the existing code in detail.
Here in this PR is what appears to be the best way to implement NPCs with the current code base: https://github.com/RafaelBarbosatec/bonfire/pull/236

This requires minimal changes, is clean, and keeps things modular with mixins. 
The current 'Enemy' class is a good starting point o create NPCs. It has all the right methods and mixins for almost everything that you'd need for most use cases.

There are couple of minor problems with this approach: 
1. The name of the base class, 'Enemy' is not right if we recommend it to be used for neutral or friendly NPCs .
2. There are some extensions which are very specific to enemies, and should be refactored into another set.

I'd suggest we address these changes in the next major version, since these will be breaking changes. The changes will include:
1. Renaming the current Enemy class to NPC. 
2. Create another class that extends NPC class. 
3. Refactor the current Enemy extensions into two- NPC extensions, and Enemy extensions.

Let me know what you think about this.

**Comment by KyleKun** [comment]
@RafaelBarbosatec I think this is a good idea, what do you think?

**Comment by RafaelBarbosatec** [comment]
Available in Version 2.4.0.

**Comment by RafaelBarbosatec** [comment]
I'm closing this issue. Any suggestions or improvements you can submit a PR or reopen this issue.
Thanks!
